File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        int n = nums.length;

        for(int i=0;i<n;i++){
            sum += nums[i];
        }

        if(sum%2!=0)return false;
        sum /=2;

        boolean[][] dp = new boolean[sum+1][n+1];
        
        Arrays.fill(dp[0], true);
        for(int i=1;i<sum+1;i++){
            for(int j=0;j<n;j++){
                if(dp[i][j])dp[i][j+1]=true;
                else if(nums[j]>i)dp[i][j+1]=false;
                else{
                    dp[i][j+1] = dp[i-nums[j]][j];
                }
            }
        }
        return dp[sum][n];
    }
}

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-0.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = static_cast<int>(nums.size());
        
        vector<int> prefix_sum(n+1,0);
        unordered_map<int,int> prefix_map;


        for(int i=0;i<n;i++){
            prefix_sum[i+1] = prefix_sum[i] + nums[i];
        }
        int count = 0;
        for(int i=0;i<n;i++){
            prefix_map[prefix_sum[i]+k]++; 
            int sum = prefix_sum[i+1];
            if(prefix_map.find(sum)!=prefix_map.end()){
                count += prefix_map[sum];
            }
        }
        return count;
    }
};
